src/blk/kernel/KernelDevice.cc

```diff
@@ -755,7 +755,6 @@ void KernelDevice::_discard_thread(uint64_t tid)
       if (thr->stop)
 	break;
       dout(20) << __func__ << " sleep" << dendl;
-      discard_cond.notify_all(); // for the thread trying to drain...
```

Comment:
But wouldn't this removal break _discard_drain(). The latter waits for the event to proceed with empty discard queue.

Reply:
Add a flag `need_notify`, set it to true before discard_drain.